src/components/structures/MessagePanel.js

```diff
@@ -0,0 +1,264 @@
+        highlightedEventId: React.PropTypes.string,
+
+        // event after which we should show a read marker
+        readMarkerEventId: React.PropTypes.string,
+
+        // event after which we should show an animating disappearance of a
+        // read marker
+        readMarkerGhostEventId: React.PropTypes.string,
+
+        // the userid of our user. This is used to suppress the read marker
+        // for pending messages.
+        ourUserId: React.PropTypes.string,
+
+        // true to suppress the date at the start of the timeline
+        suppressFirstDateSeparator: React.PropTypes.bool,
+
+        // true if updates to the event list should cause the scroll panel to
+        // scroll down when we are at the bottom of the window. See ScrollPanel
+        // for more details.
+        stickyBottom: React.PropTypes.bool,
+
+        // callback to determine if a user is the magic freeswitch conference
+        // user. Takes one parameter, which is a user id. Should return true if
+        // the user is the conference user.
+        isConferenceUser: React.PropTypes.func,
+
+        // callback which is called when the panel is scrolled.
+        onScroll: React.PropTypes.func,
+
+        // callback which is called when more content is needed.
+        onFillRequest: React.PropTypes.func,
+    },
+
+    /* get the DOM node representing the given event */
+    getNodeForEventId: function(eventId) {
+        if (!this.eventNodes) {
+            return undefined;
+        }
+
+        return this.eventNodes[eventId];
+    },
+
+    /* return true if the content is fully scrolled down right now; else false.
+     */
+    isAtBottom: function() {
+        return this.refs.scrollPanel 
+            && this.refs.scrollPanel.isAtBottom();
+    },
+
+    /* get the current scroll state. See ScrollPanel.getScrollState for
+     * details.
+     *
+     * returns null if we are not mounted.
+     */
+    getScrollState: function() {
+        if (!this.refs.scrollPanel) { return null; }
+        return this.refs.scrollPanel.getScrollState();
+    },
+
+    /* jump to the bottom of the content.
+     */
+    scrollToBottom: function() {
+        if (this.refs.scrollPanel) {
+            this.refs.scrollPanel.scrollToBottom();
+        }
+    },
+
+    /* jump to the given event id.
+     *
+     * pixelOffset gives the number of pixels between the bottom of the node
+     * and the bottom of the container. If undefined, it will put the node
+     * in the middle of the container.
+     */
+    scrollToEvent: function(eventId, pixelOffset) {
+        if (this.refs.scrollPanel) {
+            this.refs.scrollPanel.scrollToToken(eventId, pixelOffset);
+        }
+    },
+
+    /* check the scroll state and send out pagination requests if necessary.
+     */
+    checkFillState: function() {
+        if (this.refs.scrollPanel) {
+            this.refs.scrollPanel.checkFillState();
+        }
+    },
+
+    render: function() {
```

Comment:
We've loosely been sticking with what seems to be an emerging convention for react method ordering (https://reactjsnews.com/react-style-guide-patterns-i-like/) ie. render goes last, not that this is critical.